aea/test_tools/test_protocol.py

```diff
@@ -0,0 +1,169 @@
+
+    def perform_message_test(self, msg: Message) -> None:  # nosec
+        """Test message encode/decode."""
+        msg.to = "receiver"
+        assert msg._is_consistent()  # pylint: disable=protected-access
+        envelope = Envelope(to=msg.to, sender="sender", message=msg)
+        envelope_bytes = envelope.encode()
+
+        actual_envelope = Envelope.decode(envelope_bytes)
+        expected_envelope = envelope
+
+        assert expected_envelope.to == actual_envelope.to
+        assert expected_envelope.sender == actual_envelope.sender
+        assert (
+            expected_envelope.protocol_specification_id
+            == actual_envelope.protocol_specification_id
+        )
+        assert expected_envelope.message != actual_envelope.message
+
+        actual_msg = self.MESSAGE_CLASS.serializer.decode(actual_envelope.message_bytes)
+        actual_msg.to = actual_envelope.to
+        actual_msg.sender = actual_envelope.sender
+        expected_msg = msg
+        assert expected_msg == actual_msg
+
+
+class BaseProtocolMessageConstructionTestCase(BaseMessageTestCase):
+    """Base class to test message construction for the protocol."""
+
+    def test_run(self) -> None:
+        """Run the test."""
+        msg = self.build_message()
+        self.perform_message_test(msg)
+
+    @abstractmethod
+    def build_message(self) -> Message:
+        """Build the message to be used for testing."""
+
+
+class BaseProtocolMessagesTestCase(BaseMessageTestCase):
+    """Base class to test messages for the protocol."""
+
+    def test_messages_ok(self) -> None:
+        """Run messages are ok for encode and decode."""
+        for msg in self.build_messages():
+            self.perform_message_test(msg)
+
+    def test_messages_inconsistent(self) -> None:
+        """Run messages are inconsistent."""
+        for msg in self.build_inconsistent():
+            assert not msg._is_consistent()  # pylint: disable=protected-access # nosec
+
+    def test_messages_fail_to_encode_decode(self) -> None:
+        """Run messages are failing to encode and decode."""
+        msg = self.build_messages()[0]
+        with patch.object(msg.__class__.Performative, "__eq__", return_value=False):
+            with pytest.raises(
+                ValueError, match=f"Performative not valid: {msg.performative}"
+            ):
+                msg.serializer.encode(msg)
+
+        encoded_msg = msg.serializer.encode(msg)
+        with patch.object(msg.__class__.Performative, "__eq__", return_value=False):
+            with pytest.raises(
+                ValueError, match=f"Performative not valid: {msg.performative}"
+            ):
+                msg.serializer.decode(encoded_msg)
+
+    @abstractmethod
+    def build_messages(self) -> List[Message]:
+        """Build the messages to be used for testing."""
+
+    @abstractmethod
+    def build_inconsistent(self) -> List[Message]:
+        """Build inconsistent messages to be used for testing."""
```

Comment:
This is something you could scaffold. In fact, you can scaffold a message strategy to test the message construction, as you have all information you need available in the protocol definition provided by the developer. I want to stress the importance of negative cases here using an inverted strategy. Here an example for the default message. Since type annotation is a requirement, it is guaranteed to be available via `__annotations__`.
https://github.com/valory-xyz/open-aea/blob/d6c6c4d44284db7b7f7bd56bcf019bfb33999bf1/packages/valory/connections/test_libp2p/tests/test_dht.py#L245-L256

Reply:
@Karrenbelt 
agree, but it's out of scope.
already made couple of proopsals for better message structure introspection, requires to update generator

Reply:
Certainly out of scope of the PR but worth noting. Message passing is probably the most important thing to test using fuzzers, so scaffolding it here will be useful to test connections in which the protocol is used as well. Testing actual communication based on speech acts (performatives) and message content lies not only beyond the scope of this PR but also beyond that of the protocol.

Reply:
https://github.com/valory-xyz/open-aea/issues/530